src/smt.rs

```diff
@@ -1,163 +1,632 @@
+            args: vec![],
+        }
     }
 }
 
-pub enum SMTIdentifier {
-    Simple(String),
+// SMT expression builders
+
+pub fn eq(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::Eq,
+        args: vec![lhs, rhs],
+    }
 }
 
-pub enum SMTSort {
-    Simple(SMTIdentifier),
-    Parametric(SMTIdentifier, Vec<Self>),
+pub fn neq(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    not(eq(lhs, rhs))
 }
 
-impl SMTSort {
-    pub fn bitvec(index: usize) -> Self {
-        Self::Simple(SMTIdentifier::Simple(format!("(_ BitVec {})", index)))
+pub fn eq_zero(expr: SMTExpr) -> SMTExpr {
+    eq(expr, 0.into())
+}
+
+pub fn neq_zero(expr: SMTExpr) -> SMTExpr {
+    neq(expr, 0.into())
+}
+
+pub fn not(expr: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::Not,
+        args: vec![expr],
     }
 }
 
-pub enum SMTLiteral {
-    Hex(String),
+pub fn and(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::And,
+        args: vec![lhs, rhs],
+    }
 }
 
-impl SMTLiteral {
-    pub fn bv256_zero() -> Self {
-        Self::Hex("0".repeat(64))
+pub fn and_vec(args: Vec<SMTExpr>) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::And,
+        args,
     }
 }
 
-pub enum SMTExpression {
-    Eq(BoxedSMTFormat, BoxedSMTFormat),
-    Not(BoxedSMTFormat),
-    Literal(SMTLiteral),
-    Variable(SMTIdentifier),
+pub fn or(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::Or,
+        args: vec![lhs, rhs],
+    }
 }
 
-impl From<u64> for SMTExpression {
-    fn from(input: u64) -> SMTExpression {
-        SMTExpression::Literal(SMTLiteral::Hex(format!("{:064x}", input)))
+pub fn or_vec(args: Vec<SMTExpr>) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::Or,
+        args,
     }
 }
 
-impl From<SMTVariable> for SMTExpression {
-    fn from(input: SMTVariable) -> SMTExpression {
-        SMTExpression::Variable(SMTIdentifier::Simple(input.name))
+pub fn ite(cond: SMTExpr, true_term: SMTExpr, false_term: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::Ite,
+        args: vec![cond, true_term, false_term],
     }
 }
 
-pub enum SMTStatement {
-    Assert(BoxedSMTFormat),
-    DeclareConst(String, SMTSort),
-    DefineConst(String, SMTSort, String),
+pub fn bvnot(expr: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::BvNot,
+        args: vec![expr],
+    }
 }
 
-macro_rules! impl_smt {
-    (enum $type:ident {
-        $($variant:ident ( $($field:ident),* $(,)? ) $format:literal),*  $(,)?
-    }) => {
-        impl SMTFormat for $type {
-            fn as_smt(&self) -> String {
-                match self {
-                    $($type::$variant($($field),*) => format!($format, $($field.as_smt()),*),)*
-                }
-            }
+pub fn bvand(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
+    SMTExpr {
+        op: SMTOp::BvAnd,
+        args: vec![lhs, rhs],
+    }
+}
+
+pub fn bvor(lhs: SMTExpr, rhs: SMTExpr) -> SMTExpr {
```

Comment:
Can we replace all those by macros? :D

Reply:
Probably, yes. But can we maybe do it in another PR?